File: libs/sentra-shared/sentra_shared/domain/repositories/user_repository.py
from sqlalchemy.orm import Session
from sentra_shared.domain.repositories.base_repository import BaseRepository
from sentra_shared.domain.entities.user_entity import UserEntity

class UserRepository(BaseRepository[UserEntity]):
    def __init__(self, db: Session):
        super().__init__(UserEntity, db)

    
    def get_by_username(self, username: str) -> UserEntity:
        return self.db.query(UserEntity).filter(UserEntity.username == username).first()

    def get_by_email(self, email: str) -> UserEntity:
        return self.db.query(UserEntity).filter(UserEntity.email == email).first()

File: src/Sentra.Infrastructure/Sql/Repositories/UserRepository.cs
using Microsoft.EntityFrameworkCore;
using Sentra.Application.Users;
using Sentra.Domain.Entities;

namespace Sentra.Infrastructure.Sql.Repositories;

public sealed class UserRepository : IUserRepository
{
    private readonly SentraDbContext _db;

    public UserRepository(SentraDbContext db) => _db = db;

    public Task<User?> GetById(Guid id) =>
        _db.Users.FirstOrDefaultAsync(u => u.Id == id);

    public Task<User?> GetByUsername(string username) =>
        _db.Users.FirstOrDefaultAsync(u => u.Username == username);

    public Task<User?> GetByEmail(string email) =>
        _db.Users.FirstOrDefaultAsync(u => u.Email == email);

    public async Task Add(User user)
    {
        await _db.Users.AddAsync(user);
        await _db.SaveChangesAsync();
    }

    public async Task Update(User user)
    {
        _db.Users.Update(user);
        await _db.SaveChangesAsync();
    }
}
